From 2ae7fcc6de174f0741ee563e20e99db249af5ecc Mon Sep 17 00:00:00 2001
From: Kostiantyn Kostiuk <kostyanf14@live.com>
Date: Mon, 4 Aug 2025 13:24:34 +0300
Subject: [PATCH] Revert "qga: Don't daemonize before channel is initialized"

This reverts commit c6f5dd7ac8ef62dcdec4cdeda1467c658161afff.

Resolves: RHEL-107174
---
 qga/main.c | 24 ++++++------------------
 1 file changed, 6 insertions(+), 18 deletions(-)

diff --git a/qga/main.c b/qga/main.c
index 6c02f3ec38..e7fb94e973 100644
--- a/qga/main.c
+++ b/qga/main.c
@@ -1457,6 +1457,7 @@ static GAState *initialize_agent(GAConfig *config, int socket_activation)
         if (config->daemonize) {
             /* delay opening/locking of pidfile till filesystems are unfrozen */
             s->deferred_options.pid_filepath = config->pid_filepath;
+            become_daemon(NULL);
         }
         if (config->log_filepath) {
             /* delay opening the log file till filesystems are unfrozen */
@@ -1464,6 +1465,9 @@ static GAState *initialize_agent(GAConfig *config, int socket_activation)
         }
         ga_disable_logging(s);
     } else {
+        if (config->daemonize) {
+            become_daemon(config->pid_filepath);
+        }
         if (config->log_filepath) {
             FILE *log_file = ga_open_logfile(config->log_filepath);
             if (!log_file) {
@@ -1510,20 +1514,6 @@ static GAState *initialize_agent(GAConfig *config, int socket_activation)
 
     ga_apply_command_filters(s);
 
-    if (!channel_init(s, s->config->method, s->config->channel_path,
-                      s->socket_activation ? FIRST_SOCKET_ACTIVATION_FD : -1)) {
-        g_critical("failed to initialize guest agent channel");
-        return NULL;
-    }
-
-    if (config->daemonize) {
-        if (ga_is_frozen(s)) {
-            become_daemon(NULL);
-        } else {
-            become_daemon(config->pid_filepath);
-        }
-    }
-
     ga_state = s;
     return s;
 }
@@ -1562,9 +1552,8 @@ static void cleanup_agent(GAState *s)
 
 static int run_agent_once(GAState *s)
 {
-    if (!s->channel &&
-        channel_init(s, s->config->method, s->config->channel_path,
-                     s->socket_activation ? FIRST_SOCKET_ACTIVATION_FD : -1)) {
+    if (!channel_init(s, s->config->method, s->config->channel_path,
+                      s->socket_activation ? FIRST_SOCKET_ACTIVATION_FD : -1)) {
         g_critical("failed to initialize guest agent channel");
         return EXIT_FAILURE;
     }
@@ -1573,7 +1562,6 @@ static int run_agent_once(GAState *s)
 
     if (s->channel) {
         ga_channel_free(s->channel);
-        s->channel = NULL;
     }
 
     return EXIT_SUCCESS;
-- 
2.48.1

